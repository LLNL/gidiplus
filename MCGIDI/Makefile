SHELL = /bin/ksh

# <<BEGIN-copyright>>
# Copyright 2019, Lawrence Livermore National Security, LLC.
# See the top-level COPYRIGHT file for details.
# 
# SPDX-License-Identifier: MIT
# <<END-copyright>>

DIRS = Src include lib bin Doc

POPs_PATH = $(abspath ../PoPI)
SMR_PATH = $(abspath ../statusMessageReporting)
NF_PATH = $(abspath ../numericalFunctions)

.PHONY: default include lib bin clean realclean

default: include lib
	cd Test; $(MAKE) default
	$(MAKE) bin

include:
	cd include; $(MAKE)

lib:
	cd lib; $(MAKE) CXX=$(CXX) CXXFLAGS='$(CXXFLAGS)' POPs_PATH='$(POPs_PATH)' SMR_PATH='$(SMR_PATH)' NF_PATH='$(NF_PATH)'

bin:
	cd bin; $(MAKE) CXX=$(CXX) CXXFLAGS='$(CXXFLAGS)' POPs_PATH='$(POPs_PATH)' SMR_PATH='$(SMR_PATH)' NF_PATH='$(NF_PATH)' bin

check:
	cd Test; $(MAKE) CXX=$(CXX) CXXFLAGS='$(CXXFLAGS)' POPs_PATH='$(POPs_PATH)' SMR_PATH='$(SMR_PATH)' NF_PATH='$(NF_PATH)' check

speeds:
	cd Speeds; $(MAKE) CXX=$(CXX) CXXFLAGS='$(CXXFLAGS)' POPs_PATH='$(POPs_PATH)' SMR_PATH='$(SMR_PATH)' NF_PATH='$(NF_PATH)' speeds

clean:
	$(MAKE) doDIRS _DIRS="$(DIRS)" _TARGET=clean
	cd Test; $(MAKE) clean
	cd Speeds; $(MAKE) clean

realclean:
	$(MAKE) doDIRS _DIRS="$(DIRS)" _TARGET=realclean
	cd Test; $(MAKE) realclean
	cd Speeds; $(MAKE) realclean

docs:
	doxygen Doc/doxygen/Doxyfile

doDIRS:
	SAVED_PWD=`pwd`; \
	for DIR in $(_DIRS); do cd $$DIR; $(MAKE) $(_TARGET); cd $$SAVED_PWD; done
