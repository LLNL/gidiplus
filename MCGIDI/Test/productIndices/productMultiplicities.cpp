/*
# <<BEGIN-copyright>>
# Copyright 2019, Lawrence Livermore National Security, LLC.
# See the top-level COPYRIGHT file for details.
# 
# SPDX-License-Identifier: MIT
# <<END-copyright>>
*/

#include <stdlib.h>
#include <iostream>
#include <iomanip>
#include <set>

#include "MCGIDI.hpp"

#include "MCGIDI_testUtilities.hpp"

static char const *description = "Loops over temperature and energy, printing the total cross section. If projectile is a photon, see options *-a* and *-n*.";

void main2( int argc, char **argv );
/*
=========================================================
*/
int main( int argc, char **argv ) {

    try {
        main2( argc, argv ); }
    catch (std::exception &exception) {
        std::cerr << exception.what( ) << std::endl;
        exit( EXIT_FAILURE ); }
    catch (char const *str) {
        std::cerr << str << std::endl;
        exit( EXIT_FAILURE ); }
    catch (std::string &str) {
        std::cerr << str << std::endl;
        exit( EXIT_FAILURE );
    }

    exit( EXIT_SUCCESS );
}
/*
=========================================================
*/
void main2( int argc, char **argv ) {

    PoPI::Database pops;
    pops.addFile( "../../../GIDI/Test/pops.xml", false );
    GIDI::Construction::PhotoMode photo_mode = GIDI::Construction::PhotoMode::nuclearOnly;
    std::set<int> reactionsToExclude;

    std::cerr << "    " << __FILE__;
    for( int i1 = 1; i1 < argc; i1++ ) std::cerr << " " << argv[i1];
    std::cerr << std::endl;

    argvOptions2 argv_options( "crossSections", description );

    argv_options.add( argvOption2( "--map", true, "The map file to use." ) );
    argv_options.add( argvOption2( "--pid", true, "The PoPs id of the projectile." ) );
    argv_options.add( argvOption2( "--tid", true, "The PoPs id of the target." ) );
    argv_options.add( argvOption2( "-a", false, "Include photo-atomic protare if relevant. If present, disables photo-nuclear unless *-n* also present." ) );
    argv_options.add( argvOption2( "-n", false, "Include photo-nuclear protare if relevant. This is the default unless *-a* present." ) );

    argv_options.parseArgv( argc, argv );

    std::string mapFilename = argv_options.find( "--map" )->zeroOrOneOption( argv, "../../../GIDI/Test/all3T.map" );
    std::string projectileID = argv_options.find( "--pid" )->zeroOrOneOption( argv, PoPI::IDs::neutron );
    std::string targetID = argv_options.find( "--tid" )->zeroOrOneOption( argv, "O16" );

    if( argv_options.find( "-a" )->present( ) ) {
        photo_mode = GIDI::Construction::PhotoMode::atomicOnly;
        if( argv_options.find( "-n" )->present( ) ) photo_mode = GIDI::Construction::PhotoMode::nuclearAndAtomic;
    }

    GIDI::Map::Map map( mapFilename, pops );
    GIDI::Construction::Settings construction( GIDI::Construction::ParseMode::all, photo_mode );
    GIDI::Protare *protare = map.protare( construction, pops, projectileID, targetID );

    GIDI::Styles::TemperatureInfos temperatures = protare->temperatures( );
    std::string label( temperatures[0].heatedCrossSection( ) );
    MCGIDI::Transporting::MC settings( pops, projectileID, &protare->styles( ), label, GIDI::Transporting::DelayedNeutrons::on, 20.0 );

    GIDI::Transporting::Particles particles;

    GIDI::Transporting::Groups_from_bdfls groups_from_bdfls( "../../../GIDI/Test/bdfls" );
    GIDI::Transporting::Fluxes_from_bdfls fluxes_from_bdfls( "../../../GIDI/Test/bdfls", 0.0 );

    MCGIDI::DomainHash domainHash( 4000, 1e-8, 10 );
    MCGIDI::Protare *MCProtare = MCGIDI::protareFromGIDIProtare( *protare, pops, settings, particles, domainHash, temperatures, reactionsToExclude );

    MCProtare->setUserParticleIndex( pops[PoPI::IDs::neutron], 0 );
    MCProtare->setUserParticleIndex( pops["H2"], 10 );
    MCProtare->setUserParticleIndex( pops[PoPI::IDs::photon], 11 );

    for( std::size_t reactionIndex = 0; reactionIndex < MCProtare->numberOfReactions( ); ++reactionIndex ) {
        MCGIDI::Reaction const *reaction = MCProtare->reaction( reactionIndex );

        std::cout << reaction->label( ).c_str( ) << std::endl;

        auto indices = reaction->productIndices( );
        auto userIndices = reaction->userProductIndices( );
        for( MCGIDI_VectorSizeType productIndex = 0; productIndex < indices.size( ); ++productIndex ) {
            int index = indices[productIndex];

            std::cout << "    " << index << "  " << reaction->productMultiplicities( index ) << " " << userIndices[productIndex] << std::endl;
        }
    }

    delete MCProtare;

    delete protare;
}
