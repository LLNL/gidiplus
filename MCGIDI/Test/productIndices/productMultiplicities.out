n + O16
    209  1 -1
    4  1 0
n + (O16_e1 -> O16)
    209  1 -1
    4  1 0
n + (O16_e2 -> O16)
    209  1 -1
    4  1 0
n + (O16_e3 -> O16)
    209  1 -1
    4  1 0
n + (O16_e4 -> O16)
    209  1 -1
    4  1 0
n + (O16_e5 -> O16)
    209  1 -1
    4  1 0
n + (O16_e6 -> O16)
    209  1 -1
    4  1 0
n + (O16_e7 -> O16)
    209  1 -1
    4  1 0
n + (O16 -> O16 + photon) [continuum]
    209  1 -1
    4  1 0
    1  -1 11
2n + O15 + photon
    207  1 -1
    4  2 0
    1  -1 11
n + H1 + N15 + photon
    16  1 -1
    176  1 -1
    4  1 0
    1  -1 11
H1 + N16
    16  1 -1
    178  1 -1
H1 + (N16_e1 -> N16)
    16  1 -1
    178  1 -1
H1 + (N16_e2 -> N16)
    16  1 -1
    178  1 -1
H1 + (N16_e3 -> N16)
    16  1 -1
    178  1 -1
H2 + N15
    18  1 10
    176  1 -1
H2 + (N15_e1 -> N15)
    18  1 10
    176  1 -1
H2 + (N15_e2 -> N15)
    18  1 10
    176  1 -1
H2 + (N15_e3 -> N15)
    18  1 10
    176  1 -1
H2 + (N15_e4 -> N15)
    18  1 10
    176  1 -1
H2 + (N15_e5 -> N15)
    18  1 10
    176  1 -1
H2 + (N15_e6 -> N15)
    18  1 10
    176  1 -1
H2 + (N15_e7 -> N15)
    18  1 10
    176  1 -1
H2 + (N15_e8 -> N15)
    18  1 10
    176  1 -1
H3 + N14
    20  1 -1
    174  1 -1
H3 + (N14_e1 -> N14)
    20  1 -1
    174  1 -1
H3 + (N14_e2 -> N14)
    20  1 -1
    174  1 -1
He4 + C13
    143  1 -1
    35  1 -1
He4 + (C13_e1 -> C13)
    143  1 -1
    35  1 -1
He4 + (C13_e2 -> C13)
    143  1 -1
    35  1 -1
He4 + (C13_e3 -> C13)
    143  1 -1
    35  1 -1
O17 + photon
    211  1 -1
    1  -1 11
n + He4 + C12 + photon
    141  1 -1
    35  1 -1
    4  1 0
    1  -1 11
n + 4He4 + photon
    35  4 -1
    4  1 0
    1  -1 11
2He4 + Be9 + photon
    83  1 -1
    35  2 -1
    1  -1 11
